File: client/app/wordpicker/CallSignPicker.ts
import CompoundWordPicker from "./CompoundWordPicker";
import RandomWordPicker from "./RandomWordPicker";

import CallSignData from './CallSignData';

const {adjectives, nouns} = CallSignData;

export default class CallSignPicker {

    private picker: CompoundWordPicker;

    constructor() {
        const adjectivePicker = new RandomWordPicker(adjectives);
        const nounPicker = new RandomWordPicker(nouns);
        this.picker = new CompoundWordPicker([adjectivePicker, nounPicker]);
    }


    pick(email: string) {
        const [adjective, noun] = this.picker.choose(email);
        return {
            adjective: adjective,
            noun: noun
        }
    }
}

File: commonKt/src/commonMain/kotlin/com/zegreatrob/coupling/common/entity/callsign/CallSignPicker.kt
package com.zegreatrob.coupling.common.entity.callsign

import kotlin.random.Random


interface CallSignPicker : PredictableWordPicker {

    fun AvailableComponents.pick(email: String) = CallSign(
            shuffledAdjectives().pickForGiven(email),
            shuffledNouns().pickForGiven(email)
    )

    private val adjectiveRandom get() = Random(0)
    private val nounRandom get() = Random(1)

    private fun AvailableComponents.shuffledAdjectives() = adjectives.shuffled(adjectiveRandom)
    private fun AvailableComponents.shuffledNouns() = nouns.shuffled(nounRandom)
}

data class AvailableComponents(val adjectives: List<String>, val nouns: List<String>)

data class CallSign(val adjective: String, val noun: String)